biz.aQute.bndlib/src/aQute/bnd/plugin/module/ModuleInfoPlugin.java

```diff
@@ -0,0 +1,399 @@
+				writer.requires(moduleToAdd, 0, null);
+
+				requiredModules.add(moduleToAdd);
+			}
+		}
+
+		externallyReferred.entrySet()
+			.stream()
+			.filter(entry -> {
+				PackageRef packageRef = entry.getKey();
+				Attrs attrs = index.get(packageRef);
+
+				if (attrs == null || !attrs.containsKey(INTERNAL_MODULE_DIRECTIVE)) {
+					String eeModuleName = moduleEE.getModules()
+						.entrySet()
+						.stream()
+						.filter(it -> it.getValue()
+							.getTyped(Attrs.LIST_STRING, "exports")
+							.contains(packageRef.getFQN()))
+						.map(Entry::getKey)
+						.findAny()
+						.orElse(null);
+
+					if (eeModuleName != null) {
+						index.put(packageRef, Attrs.create(INTERNAL_MODULE_DIRECTIVE, eeModuleName));
+						return true;
+					}
+
+					if (logger.isWarnEnabled())
+						logger.warn("Can't find a module name for imported package: {}", packageRef.getFQN());
+
+					return false;
+				}
+				return true;
+			})
+			// group packages by module/contract
+			.collect(groupingBy(entry -> index.get(entry.getKey())
+				.get(INTERNAL_MODULE_DIRECTIVE)))
+			.entrySet()
+			.forEach(entry -> {
+				String moduleName = entry.getKey();
+
+				if (requiredModules.contains(moduleName)) {
+					return;
+				}
+
+				List<Entry<PackageRef, Attrs>> referencedPackages = entry.getValue();
+
+				// An import results in `transitive` requires where there is an
+				// `Export-Package` that has a `uses` constraint on the imported
+				// package.
+				boolean isTransitive = exports.entrySet()
+					.stream()
+					.map(export -> export.getValue()
+						.get(USES_DIRECTIVE))
+					.flatMap(uses -> Strings.split(uses)
+						.stream())
+					.map(use -> analyzer.getPackageRef(Descriptors.fqnToBinary(use)))
+					.anyMatch(packageRef -> referencedPackages.stream()
+						.anyMatch(e -> e.getKey()
+							.equals(packageRef)));
+
+				// TODO modules can fall under the follow categories:
+				// a) JDK modules (whose packages are not _yet_ imported)
+				// b) all packages referenced are `resolution:=optional` or
+				// `Dynamic-ImportPackage`
+				// c) statically referenced classes (which do not incur an
+				// import like bundle annotations)
+				// b) and c) result in static requires
+				boolean isStatic = false;
+
+				// Allowed: 0 | ACC_TRANSITIVE | ACC_STATIC_PHASE |
+				// ACC_SYNTHETIC | ACC_MANDATED
+				int access = (isTransitive ? ModuleAttribute.Require.ACC_TRANSITIVE : 0)
+					| (isStatic ? ModuleAttribute.Require.ACC_STATIC_PHASE : 0);
+
+				if (moduleName.equals(MODULE_JAVA_BASE)) {
+					access |= ModuleAttribute.Require.ACC_MANDATED;
+				}
+
+				// TODO collect module version. Do we want module version? It is
+				// not checked at runtime.
+
+				writer.requires(moduleName, access, null);
+
+				requiredModules.add(moduleName);
+			});
+
+		if (!requiredModules.contains(MODULE_JAVA_BASE)) {
+			writer.requires(MODULE_JAVA_BASE, ModuleAttribute.Require.ACC_MANDATED, null);
```

Comment:
You do not need to add `java.base`. It is always added by ModuleInfoWriter. 

https://github.com/bndtools/bnd/blob/e94ac48cdcef4608318513502ee1cb0d0a17bb93/biz.aQute.bndlib/src/aQute/bnd/classfile/writer/ModuleInfoWriter.java#L37